docs/about/coding_guidelines.md

```diff
@@ -0,0 +1,68 @@
+refer to them once in a while when in doubt.
+
+
+### Defining classes and traits
+
+1. Value classes must be final and extend AnyVal.
+2. Method extension classes must be final and extend `AnyVal`.
+3. Sealed traits that are ADTs should extend `Product` and `Serializable`.
+4. Regular traits and sealed trait that do not form ADTS should extend `Serializable` but not `Product`.
+5. Multiple parameter classes that are ADT should be case class and not extend AnyVal
+6. Traits should always extends `Serializable`. (i.e. `ZIO`)
+
+### Final and private modifiers 
+
+1. All methods on classes / traits are declared `final`, by default.
+2. No methods on objects declared `final`.
+3. No methods on final classes declared `final`.
+4. All classes inside objects should be defined `final`.
+5. All final classes have private constructors & constructor parameters.
+6. All `vals` declared `final`, either in objects or `final classes`, if they are constant expressions and without type annotations.
+7. Private `methods` and `vals` are not declared `final` since they are already final.
+8. Package-private `vals` and `methods` declared `final`.
+
+### Refactoring
+
+1. If a class has all `final` members, the class should be declared `final` and `final` member annotations should be removed except constant expressions.
+2. All type annotations should use the least powerful type alias. This means, that, let us say, a `ZIO` effect that has no dependencies but throws an arbitrary error, should be defined as `IO`.
+
+### Naming of parameters in methods and classes
+
+1. Partial functions are called `pf`
+2. Evidences are called `ev`
+3. Promise are called `p` (unless in its own class methods, in that case it is called `that`)
+4. Functions are called `fn`, `fn1`, unless they bear specific meaning: `use`, `release`, ..
+4. Consider _ method having more meaningful names
+5. Iterable are called `in`
+6. When a parameter type equals own (in a method of a trait) call it `that`
+7. Never use by-name parameter unless you really know what you are doing and it makes sense. Mind the `Function[0]` extra allocation and loss of clean syntax when invoking the method.
+8. Folds initial value are called `zero`
+
+### Misc
+
+1. GADTS should always have type annotation.
+2. Type alias should always have type annotation.
+3. Methods that lift pure values to effects are dangerous. Dangerous in the sense that they can potentially have dangerous side-effects. 
+Such methods should have a default lazy variant and an eager variant for advanced users that are aware they absolutely do not have side-effects in their code, 
+having slight gains in performance. The lazy variant should have a normal name (succeed, fail, lift, potato) and the eager variant should have a now suffix 
+(succeedNow, liftNow, potatoNow) which makes it clear of its eager behaviour.      
+
+### Code structure
+
+1. Method alphabetization - add lazy `vals` to fix forward references.
```

Comment:
Would be good to justify these and spell them out in a bit more detail.

For example, generally we have:

1. Public abstract defs / vals listed first, and alphabetized, with operators appearing before names.
2. Public concrete defs / vals listed second, and alphabetized, with operators appearing before names.
3. Private implementation details listed third, and alphabetized, with operators appearing before names.

Would be nice to enforce using Scalafix ultimately.

Reply:
@jdegoes Is there any rule of thumb regarding when to make something a val or a def? Something that could be somehow formalized in a sentence. 

It comes to mind when a colleague approached me to ask about clock live instance:

`val nanoTime: UIO[Long] = IO.effectTotal(System.nanoTime)`

```
def currentDateTime: ZIO[Any, Nothing, OffsetDateTime] = 
     for {
          millis <- currentTime(TimeUnit.MILLISECONDS)
          zone   <- ZIO.effectTotal(ZoneId.systemDefault)
        } yield OffsetDateTime.ofInstant(Instant.ofEpochMilli(millis), zone)
```

The explanation I could come up with was that nanoTime was just a pure value (effect) whereas currentDateTime was composing several. But does that really matter? it is still a value and it takes no parameters. 

Additionally, I considered whether it mattered how flexible `nanoTime` and `currentDateTime` should be when thinking of overriding them.

Reply:
Ultimately I am hoping to pick up kind of an epic where I focus on scalafix rules (dream of zio-lint) but we will see how that works

Reply:
@saraiva132 A general guideline would be that if something has no parameters and is purely functional it should be a `val` instead of a `def`. In your example above I don't think there is a good reason for that to be a `def` instead of a `val`. Reasons to not do that could include avoiding early initialization issues, preventing forward reference errors, or if although the value is pure it is computationally expensive and may not be needed.

Reply:
@adamfraser agreed. I was thinking similarly. In general, to avoid forward references a lazy val would be preferred, I believe. Regarding the computationally expensive since these are only descriptions of programs and nothing is executed I think using 
values might be a good practice also as a mental model ( people using effects as values) . I start to see some things that could potentially be made into Scala fix rule,